### PR TITLE
Refactor: SQL write and read path, split SQL tests

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
This PR does refactoring on:

- [x]  Splits sql_ingest.go into multiple files (1st commit)
- [x] Shift the ingestor component into its respective dir (2nd commit)
- [x]  Split SQL reader and writer tests
- [x]  Shift querier component into its respective dir
- [x]  License headers across files in the project
- [x]  Shift minTime and maxTime to utils so as to keep a common source for api/common.go and delete.go

### CODE REVIEW

File: pkg/api/delete.go [hunk 1]
```diff
@@ -2,13 +2,14 @@ package api
 
 import (
 	"fmt"
+	delete2 "github.com/timescale/promscale/pkg/pgmodel/delete"
```
Comment (antekresic) [inline pkg/api/delete.go:5]: This name `delete2` sticks out like a sore thumb. 
Makes the code confusing, personally I would need to try to find out what the `2` stands for.

```suggestion
	deletePkg "github.com/timescale/promscale/pkg/pgmodel/delete"
```
Also, imports need to be properly grouped and sorted.

File: pkg/api/query_test.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/timescale/promscale/pkg/pgmodel/querier"
```
Comment (antekresic) [inline pkg/api/query_test.go:8]: Import needs to be grouped with other Promscale packages.

File: pkg/pgclient/client.go [hunk 3]
```diff
@@ -3,6 +3,10 @@ package pgclient
 import (
 	"context"
 	"fmt"
+	"github.com/timescale/promscale/pkg/pgmodel/cache"
+	"github.com/timescale/promscale/pkg/pgmodel/ingestor"
+	querier2 "github.com/timescale/promscale/pkg/pgmodel/querier"
```
Comment (antekresic) [inline pkg/pgclient/client.go:8]: Is there a good reason for `querier2` naming? I'd rather change the name of the local variable its clashing with instead of the import.

File: pkg/pgmodel/ingestor/ingestor.go [hunk 4]
```diff
@@ -1,56 +1,42 @@
-// This file and its contents are licensed under the Apache License 2.0.
-// Please see the included NOTICE for copyright information and
-// LICENSE for a copy of the license.
-
-package pgmodel
+package ingestor
```
Comment (antekresic) [inline pkg/pgmodel/ingestor/ingestor.go:1]: Missing copyright header.

File: pkg/query/queryable.go [hunk 5]
```diff
@@ -2,21 +2,22 @@ package query
 
 import (
 	"context"
+	querier2 "github.com/timescale/promscale/pkg/pgmodel/querier"
```
Comment (antekresic) [inline pkg/query/queryable.go:5]: Reason for name `querier2`?

File: pkg/tests/end_to_end_tests/create_test.go [hunk 6]
```diff
@@ -15,7 +15,8 @@ import (
 	pgx "github.com/jackc/pgx/v4"
 	"github.com/jackc/pgx/v4/pgxpool"
 	_ "github.com/jackc/pgx/v4/stdlib"
-	. "github.com/timescale/promscale/pkg/pgmodel"
+	igstr "github.com/timescale/promscale/pkg/pgmodel/ingestor"
```
Comment (antekresic) [inline pkg/tests/end_to_end_tests/create_test.go:18]: Again, I prefer to change variable names if they are clashing to package names. 
Also, we are using `ingr`, `ingstr`,  and here `igstr` for a short version of the word `ingestor`. This needs to be consistent across the project.

**Review by antekresic — CHANGES_REQUESTED** [review]
Lots of things are pushed into `utils` package as a catch-all. That doesn't improve the situation from before where we had all those things in `pgmodel`, just adds more imports everywhere. 

These things need to be scoped, grouped and moved into their logical packages IMO.

Lots of missing copyright headers in new files (one or two places in existing files too). 
Some questionable import renaming that need to be addressed.

File: pkg/api/delete.go (hunk 1 above)
Comment (Harkishen-Singh) [reply pkg/api/delete.go:5]: Sorry, that was by the code editor. I will fix all these.

File: pkg/pgmodel/ingestor/ingestor.go (hunk 4 above)
Comment (Harkishen-Singh) [reply pkg/pgmodel/ingestor/ingestor.go:1]: BTW, do we need to have the copyright header across each and every file in the project? (Just for my info)

File: pkg/query/queryable.go (hunk 5 above)
Comment (Harkishen-Singh) [reply pkg/query/queryable.go:5]: Sorry, that's by the editor. Will change these in the next commit.

File: pkg/pgmodel/migrate.go [hunk 7]
```diff
@@ -390,14 +391,14 @@ func replaceSchemaNames(r io.ReadCloser) (string, error) {
...
 	s := buf.String()
-	s = strings.ReplaceAll(s, "SCHEMA_CATALOG", catalogSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_EXT", extSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_PROM", promSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_SERIES", seriesViewSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_METRIC", metricViewSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_DATA_SERIES", dataSeriesSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_DATA", dataSchema)
-	s = strings.ReplaceAll(s, "SCHEMA_INFO", infoSchema)
+	s = strings.ReplaceAll(s, "SCHEMA_CATALOG", utils.CatalogSchema)
```
Comment (antekresic) [inline pkg/pgmodel/migrate.go:394]: These things don't look like they should be in `utils` package... 
How about we make a new package called `common`? Might make more sense IMO.

File: pkg/pgmodel/utils/utils.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
+// This file and its contents are licensed under the Apache License 2.0.
```
Comment (antekresic) [inline pkg/pgmodel/utils/utils.go:1]: This might be another candidate for a `common` package instead of `utils`.

**Review by antekresic — CHANGES_REQUESTED** [review]
Looks better. Still some minor things to sort out, after that should be good IMO.

File: pkg/pgmodel/ingestor/ingestor_test.go [hunk 9]
```diff
@@ -1,95 +1,16 @@
...
+
+package ingestor
 
 import (
 	"fmt"
-	"testing"
-
+	"github.com/timescale/promscale/pkg/pgmodel/utils"
 	"github.com/timescale/promscale/pkg/prompb"
+	"testing"
```
[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/pgmodel/ingestor/ingestor_test.go:11]: Agreed. I keep missing this for some reason because the auto import in goland is kind of causing difficulties.

File: pkg/api/health.go [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
 import (
+	"github.com/timescale/promscale/pkg/pgmodel/heath"
 	"net/http"
 
 	"github.com/timescale/promscale/pkg/log"
-	"github.com/timescale/promscale/pkg/pgmodel"
 )
 
-func Health(hc pgmodel.HealthCheckerFn) http.HandlerFunc {
+func Health(hc heath.HealthCheckerFn) http.HandlerFunc {
```
Comment (blagojts) [inline pkg/api/health.go:14]:
```suggestion
func Health(hc health.HealthCheckerFn) http.HandlerFunc {
```

File: pkg/pgmodel/ingestor/buffer.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	// was chosen arbitrarily.
+	flushSize                = 2000
+	getCreateMetricsTableSQL = "SELECT table_name FROM " + schema.Catalog + ".get_or_create_metric_table_name($1)"
+	finalizeMetricCreation   = "CALL " + schema.Catalog + ".finalize_metric_creation()"
+	getSeriesIDForLabelSQL   = "SELECT * FROM " + schema.Catalog + ".get_or_create_series_id_for_kv_array($1, $2, $3)"
+	getEpochSQL              = "SELECT current_epoch FROM " + schema.Catalog + ".ids_epoch LIMIT 1"
+	maxCopyRequestsPerTxn    = 100
+)
+
+// epoch for the ID caches, -1 means that the epoch was not set
```
Comment (blagojts) [inline pkg/pgmodel/ingestor/buffer.go:27]: Do you think it's good to have a `const EpochNotSet = Epoch(-1)` so we can use it as a const throughout the code?

File: pkg/pgmodel/querier/series_set.go [hunk 12]
```diff
@@ -1,14 +1,20 @@
-package pgmodel
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package querier
 
 import (
 	"fmt"
+	"github.com/timescale/promscale/pkg/pgmodel/common/errors"
```
Comment (blagojts) [inline pkg/pgmodel/querier/series_set.go:9]: auto import

File: pkg/tests/end_to_end_tests/drop_test.go [hunk 13]
```diff
@@ -25,7 +26,7 @@ func TestSQLRetentionPeriod(t *testing.T) {
 		ts := []prompb.TimeSeries{
 			{
 				Labels: []prompb.Label{
-					{Name: MetricNameLabelName, Value: "TEST"},
+					{Name: utils.MetricNameLabelName, Value: "test"},
```
Comment (blagojts) [inline pkg/tests/end_to_end_tests/drop_test.go:29]: is the "TEST" reduced to lower case on purpose?

File: pkg/tests/end_to_end_tests/migrate_test.go [hunk 14]
```diff
@@ -137,15 +138,15 @@ func TestMigrateTwice(t *testing.T) {
...
 	testhelpers.WithDB(t, *testDatabase, testhelpers.NoSuperuser, func(db *pgxpool.Pool, t testing.TB, connectURL string) {
 		performMigrate(t, connectURL, testhelpers.PgConnectURL(*testDatabase, testhelpers.Superuser))
-		if *useExtension && !pgmodel.ExtensionIsInstalled {
+		if *useExtension && !utils.ExtensionIsInstalled {
 			t.Errorf("extension is not installed, expected it to be installed")
 		}
 
 		//reset the flag to make sure it's set correctly again.
-		pgmodel.ExtensionIsInstalled = false
+		utils.ExtensionIsInstalled = false
```
Comment (blagojts) [inline pkg/tests/end_to_end_tests/migrate_test.go:146]: man this feels wrong. I really really don't like global state. But I would leave this for another PR

**Review by blagojts — CHANGES_REQUESTED** [review]
I love the new changes, especially that the pgmodel 
package is no longer huge. 🔥 

But I dislike the utils package. 'Utils' in general is 
a code smell, since it provides no real information. 
I would suggest we move some of the things currently there 
in the `pgmodel/common` package
or a new `pgmodel/model` package (and rename `pgmodel`). 
More specifically the model package should contain:
 - labels.go
 - labels_reader.go
 - samples.go
 - MetricNameLabelName, MinTime, MaxTime consts
 - SeriesId as a type 
since these all are very data model related, they are
not just utilities.
`utils/extension` should go to `common` and I agree with
Ante that utils.go should go into `common`

**Comment by cevian** [comment]
I am taking myself and Josh off this review. Will let Ante and Blagoj shepherd this one.

File: pkg/tests/end_to_end_tests/drop_test.go (hunk 13 above)
Comment (Harkishen-Singh) [reply pkg/tests/end_to_end_tests/drop_test.go:29]: I guess, it happened while resolving conflicts.

File: pkg/pgmodel/ingestor/buffer.go (hunk 11 above)
Comment (Harkishen-Singh) [reply pkg/pgmodel/ingestor/buffer.go:27]: I am not sure. The changes as such are huge. Maybe, a follow on PR (by me or someone else who has more used to epoch code) for this and the comment below for global state?

File: pkg/pgmodel/common/extension/extension.go [hunk 15]
```diff
@@ -2,7 +2,7 @@
 // Please see the included NOTICE for copyright information and
 // LICENSE for a copy of the license.
 
-package pgmodel
+package extension
```
Comment (blagojts) [inline pkg/pgmodel/common/extension/extension.go:5]: ❤️

File: pkg/pgmodel/ingestor/buffer.go (hunk 11 above)
Comment (blagojts) [reply pkg/pgmodel/ingestor/buffer.go:27]: 👍

**Review by blagojts — APPROVED** [review]

**Review by antekresic — APPROVED** [review]
👍
